pkg/infrastructure/services/prometheus.go

```diff
@@ -18,66 +18,151 @@ import (
 	monv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
 	"github.com/kiegroup/kogito-cloud-operator/pkg/apis/app/v1alpha1"
 	"github.com/kiegroup/kogito-cloud-operator/pkg/client"
+	"github.com/kiegroup/kogito-cloud-operator/pkg/client/kubernetes"
 	"github.com/kiegroup/kogito-cloud-operator/pkg/framework"
+	"github.com/kiegroup/kogito-cloud-operator/pkg/infrastructure"
 	"github.com/kiegroup/kogito-cloud-operator/pkg/operator"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/intstr"
+	"net/http"
+	"time"
 )
 
 const prometheusServerGroup = "monitoring.coreos.com"
 
-// IsPrometheusAvailable checks if Prometheus CRD is available in the cluster
-func IsPrometheusAvailable(client *client.Client) bool {
+func configurePrometheus(client *client.Client, kogitoService v1alpha1.KogitoService, scheme *runtime.Scheme) (reconcileAfter time.Duration, err error) {
+	prometheusAvailable := isPrometheusAvailable(client)
+	if !prometheusAvailable {
+		log.Debugf("prometheus operator not available in namespace")
+		return
+	}
+	prometheusAddOnAvailable, err := isPrometheusAddOnAvailable(client, kogitoService)
+	if err != nil {
+		reconcileAfter = time.Second * 10
+		return
+	}
+	if prometheusAddOnAvailable {
+		if err = createPrometheusServiceMonitorIfNotExists(client, kogitoService, scheme); err != nil {
+			return
+		}
+	}
+	return
+}
+
+// isPrometheusAvailable checks if Prometheus CRD is available in the cluster
+func isPrometheusAvailable(client *client.Client) bool {
 	return client.HasServerGroup(prometheusServerGroup)
 }
 
-// CreateServiceMonitor create ServiceMonitor used for scraping by prometheus for kogito service
-func CreateServiceMonitor(kogitoRuntime *v1alpha1.KogitoRuntime) *monv1.ServiceMonitor {
-	monitoring := kogitoRuntime.Spec.Monitoring
-	if monitoring.Scrape {
-		endPoint := monv1.Endpoint{}
-		endPoint.TargetPort = &intstr.IntOrString{IntVal: getServiceHTTPPort(kogitoRuntime)}
-
-		if len(monitoring.Path) > 0 {
-			endPoint.Path = monitoring.Path
-		} else {
-			endPoint.Path = v1alpha1.MonitoringDefaultPath
-		}
+func isPrometheusAddOnAvailable(client *client.Client, kogitoService v1alpha1.KogitoService) (bool, error) {
+	url, err := infrastructure.FetchKogitoServiceURI(client, kogitoService.GetName(), kogitoService.GetNamespace())
+	if err != nil {
+		return false, err
+	}
+
+	url = url + getMonitoringPath(kogitoService.GetSpec().GetMonitoring())
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return false, err
+	}
+	if resp, err := client.HTTPCli.Do(req); err != nil {
```

Comment:
Here, instead of performing the request directly we can wait for the Deployment status of the service, if it's ready to receive connections we do it, otherwise just return without error and try in the next reconciliation loop.

The Deployment `status.availableReplicas` > 0 will tell you if it's ready to receive requests. This way we won't keep making failed requests to the deployed service.

Reply:
done